yt/yt/server/http_proxy/component_discovery.cpp

```diff
@@ -0,0 +1,408 @@
+
+        if (instance.Online && (!version || !startTime)) {
+            instance.Error = TError("Component is missing some of the required attributes in response")
+                << TErrorAttribute("version", version)
+                << TErrorAttribute("start_time", startTime);
+        }
+
+        instances.push_back(instance);
+    }
+
+    return instances;
+}
+
+std::vector<TYtInstance> TComponentDiscoverer::ListProxies(EYtComponentType component) const
+{
+    TGetNodeOptions options;
+    FillMasterReadOptions(options, MasterReadOptions_);
+
+    switch (component) {
+        case EYtComponentType::RpcProxies:
+            options.Attributes = {
+                "start_time",
+                "version",
+                "banned",
+            };
+            break;
+        case EYtComponentType::HttpProxies:
+            options.Attributes = {
+                "liveness",
+                "start_time",
+                "version",
+                "banned",
+            };
+            break;
+        default:
+            YT_ABORT();
+    }
+
+    auto nodeYson = WaitFor(Client_->GetNode(GetCypressDirectory(component), options))
+        .ValueOrThrow();
+
+    std::vector<TYtInstance> instances;
+    auto timeNow = TInstant::Now();
+    for (const auto& [address, node] : ConvertTo<THashMap<TString, IMapNodePtr>>(nodeYson)) {
+        auto version = node->Attributes().Find<TString>("version");
+        auto banned = node->Attributes().Find<bool>("banned");
+        auto startTime = node->Attributes().Find<TString>("start_time");
+
+        TYtInstance instance;
+        instance.Type = GetInstanceType(component);
+        instance.Address = address;
+
+        if (version && startTime) {
+            instance.Version = *version;
+            instance.StartTime = *startTime;
+            instance.Banned = banned.value_or(false);
+        } else {
+            instance.Error = TError("Cannot find required attributes in response")
+                << TErrorAttribute("version", version)
+                << TErrorAttribute("start_time", startTime);
+        }
+
+        if (component == EYtComponentType::RpcProxies) {
+            auto alive = node->AsMap()->FindChild("alive");
+            instance.Online = static_cast<bool>(alive);
+        } else {
+            auto livenessPtr = node->Attributes().Find<TLivenessPtr>("liveness");
+            if (!livenessPtr) {
+                instance.Error = TError("Liveness attribute is missing");
+            } else {
+                instance.Online = (livenessPtr->UpdatedAt + ProxyDeathAge_ >= timeNow);
+            }
+        }
+        if (instance.Online) {
+            instance.State = "online";
+        } else {
+            instance.State = "offline";
+        }
+        instances.push_back(instance);
+    }
+
+    return instances;
+}
+
+std::vector<TString> TComponentDiscoverer::GetCypressSubpaths(EYtComponentType component) const
+{
+    std::vector<TString> paths;
```

Comment:
nip: reserve below?

Reply:
Honestly, it's a little annoying to do in this case and it's typically a short list anyway, so it should fit into the default-allocated capacity.